experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts

```diff
@@ -172,4 +173,100 @@ describe("Fuzz - Targeted", () => {
 			emitter,
 		});
 	});
+
+	const undoRedoWeights: Partial<EditGeneratorOpWeights> = {
+		insert:1,
+		delete:1,
+		start:0,
+		commit: 0
+	};
+
+	describe.only("Inorder undo/redo matches the initial/final state", () => {
+		const generatorFactory = (): AsyncGenerator<TreeOperation, UndoRedoFuzzTestState> =>
+			takeAsync(opsPerRun, makeOpGenerator(undoRedoWeights));
+
+		const model: DDSFuzzModel<
+			SharedTreeTestFactory,
+			Operation,
+			DDSFuzzTestState<SharedTreeTestFactory>
+		> = {
+			workloadName: "SharedTree",
+			factory: new SharedTreeTestFactory(onCreate),
+			generatorFactory,
+			reducer: fuzzReducer,
+			validateConsistency: () => {},
+		};
+		const emitter = new TypedEventEmitter<DDSFuzzHarnessEvents>();
+		emitter.on("testStart", (initialState: UndoRedoFuzzTestState) => {
+			initialState.initialTreeState = toJsonableTree(initialState.clients[0].channel);
+			initialState.firstAnchors = []
+			// creates an initial anchor for each tree
+			for (const client of initialState.clients){
+				initialState.firstAnchors.push(getFirstAnchor(client.channel))
+			}
+		});
+		emitter.on("testEnd", (finalState: UndoRedoFuzzTestState) => {
+			const clients = finalState.clients
+
+			const finalTreeStates = []
+			// undo all of the changes and validate against initialTreeState for each tree
+			for (const [i, client] of clients.entries()){
+				const tree = client.channel;
+
+				// save final tree states to validate redo later
+				finalTreeStates.push(toJsonableTree(tree))
+
+				// call undo() until tree no more edits left
+				for (let j = 0; j<opsPerRun; j++){
+					tree.undo()
+				}
+				assert(finalState.initialTreeState !== undefined);
+				validateTree(client.channel, finalState.initialTreeState)
```

Comment:
don't you need to synchronize client state before all trees will have arrived back at their original value?

Reply:
synchronized after calling `undo()` on all clients